client/src/lib.rs

```diff
@@ -1,45 +1,356 @@
-use simperby_common::*;
+use simperby_core::*;
+use simperby_settlement::execution::convert_transaction_to_execution;
 use simperby_settlement::*;
+use std::collections::HashMap;
+use std::{sync::Arc, time::Duration};
+use rust_decimal::prelude::ToPrimitive;
 
-pub struct Ethereum {}
+const HEX_ADDRESS_LEN_WITH_PREFIX: usize = 42;
+
+abigen!(
+    ITreasury,
+    r#"[
+        function name() external view returns (string memory)
+        function chainName() external view returns (bytes memory)
+        function contractSequence() external view returns (uint128)
+        function lightClient() external view returns (uint64 heightOffset, bytes memory lastHeader)
+        function viewCommitRoots() external view returns (bytes32[] memory commitRoots)
+        function updateLightClient(bytes memory header, bytes memory proof) public
+        function execute(bytes memory transaction,bytes memory executionHash, uint64 blockHeight, bytes memory merkleProof) public
+    ]"#,
+);
+
+abigen!(
+    IERC20,
+    r#"[
+        function balanceOf(address account) external view returns (uint256)
+        function totalSupply() public view returns (uint256)
+        function transfer(address _to, uint256 _value) public returns (bool success)
+        function transferFrom(address _from, address _to, uint256 _value) public returns (bool success)
+    ]"#,
+);
+
+abigen!(
+    IERC721,
+    r#"[
+        function balanceOf(address account) external view returns (uint256)
+        function tokenOfOwnerByIndex(address owner, uint256 index) external view returns (uint256)
+    ]"#,
+);
+
+pub struct ChainConfigs {
+    /// The RPC URL of the chain
+    rpc_url: String,
+    /// The name of the chain
+    chain_name: Option<String>,
+}
+
+pub enum ChainType {
+    Ethereum(ChainConfigs),
+    Goerli(ChainConfigs),
+    Other(ChainConfigs),
+}
+
+impl ChainType {
+    fn get_rpc_url(&self) -> &str {
+        match self {
+            ChainType::Ethereum(chain) => chain.rpc_url.as_str(),
+            ChainType::Goerli(chain) => chain.rpc_url.as_str(),
+            ChainType::Other(chain) => chain.rpc_url.as_str(),
+        }
+    }
+
+    fn get_chain_name(&self) -> &str {
+        match self {
+            ChainType::Ethereum(_) => "Ethereum",
+            ChainType::Goerli(_) => "Goerli",
+            ChainType::Other(configs) => {
+                if configs.chain_name.is_some() {
+                    configs.chain_name.as_ref().unwrap().as_str()
+                } else {
+                    "Unknown"
+                }
+            }
+        }
+    }
+}
+
+pub struct Treasury {
+    /// The address of the treasury
+    pub address: HexSerializedVec,
+    /// The symbol or the name of fungible token to the address of the token
+    pub fungible_tokens: Option<HashMap<String, HexSerializedVec>>,
+    /// The symbol or the name of non-fungible token to the address of the token
+    pub non_fungible_tokens: Option<HashMap<String, HexSerializedVec>>,
+}
```

Comment:
Please remove this struct.
1. It is not used currently, and it will not be I guess.
2. Having a list of well-known tokens (e.g., USDC and USDT) is very useful, but it does not belong to here. Rather the user of `SettlementChain` would need the information.
3. The methods seem pointless. It's just complicated wrapper of  accessing the `Option<HashMap>`

Reply:
fixed